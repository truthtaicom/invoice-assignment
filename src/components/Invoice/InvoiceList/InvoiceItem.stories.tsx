import { storiesOf } from '@storybook/react';
import React from 'react';
import { action } from '@storybook/addon-actions';
import InvoiceList from "./InvoiceList"
import { simpleInvoiceListData } from './InvoiceList.utils'
import { omitDateFieldSelector } from './InvoiceList.selector';

storiesOf("Component/Invoice List", module)
  .add("List", () => (
    <InvoiceList
      deleteInvoice={action("deleteInvoice")}
      editInvoice={action("editInvoice")}
      addInvoice={action("addInvoice")}
      selectItem={action("selectItem")}
      data={simpleInvoiceListData}
      getInvoices={action("getInvoices")}
      selectedItem={{}}
      setModeInvoice={action('setModeInvoice')}
      setTabActive={action('setTabActive')}
      resetState={action('resetState')}
      setRetrieveFromBankAcc={action('setRetrieveFromBankAcc')}
      selectPaymentItem={action('action')}
      selectedPaymentItem={{}}
    />
  ))
  // .add("List without Control", () => (
  //   <InvoiceList
  //     deleteInvoice={action("deleteInvoice")}
  //     editInvoice={action("editInvoice")}
  //     addInvoice={action("addInvoice")}
  //     selectItem={action("selectItem")}
  //     data={simpleInvoiceListData}
  //     getInvoices={action("getInvoices")}
  //     selectedItem={{}}
  //   />
  // ))
  // .add("List without Control & Date", () => (
  //   <InvoiceList
  //     deleteInvoice={action("deleteInvoice")}
  //     editInvoice={action("editInvoice")}
  //     addInvoice={action("addInvoice")}
  //     selectItem={action("selectItem")}
  //     data={omitDateFieldSelector({
  //       invoices: {
  //         data: simpleInvoiceListData
  //       }
  //     })}
  //     getInvoices={action("getInvoices")}
  //     selectedItem={{}}
  //   />
  // ))
  // .add("Playground", () => (
  //   <InvoiceList
  //     deleteInvoice={action("deleteInvoice")}
  //     editInvoice={action("editInvoice")}
  //     addInvoice={action("addInvoice")}
  //     selectItem={action("selectItem")}
  //     data={simpleInvoiceListData}
  //     getInvoices={action("getInvoices")}
  //     selectedItem={{}}
  //   />
  // ))