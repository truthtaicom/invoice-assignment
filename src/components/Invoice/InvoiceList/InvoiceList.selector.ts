import { createSelector } from 'reselect'

const invoiceListSelector = state => state.invoicesReducer.data

export const omitDateFieldSelector = createSelector(
  invoiceListSelector,
  (items) => items.map(({ date, ...item }) => ({ ...item }))
)