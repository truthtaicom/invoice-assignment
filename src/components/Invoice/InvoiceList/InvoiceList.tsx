import React, { useState, useEffect } from 'react'
import styled from 'styled-components';
import InvoiceItem from './InvoiceItem'
import { SimpleList, Text, Button, Modal, Tabs, TabItem, Loading } from '../../../ui-kit'
import InvoiceInformation from '../InvoiceInformation/InvoiceInformation'
import InvoicePayment from '../InvoicePayment/InvoicePayment'
import { IInvoiceItem } from './InvoiceItem'

const StyledInvoiceList = styled.div`
  display: grid;
  grid-template-rows: 5rem auto;
`

const StyledInvoiceListHead = styled.div`
  display: grid;
  grid-template-columns: auto auto;
  padding: 1rem;
`

const StyledInvoiceListTitle = styled.div``

const StyledInvoiceListAddBtn = styled.div`
  text-align: right;
`

const StyledInvoiceListBody = styled.div`
  padding: 1rem;
`

const StyledModalConfirm = styled.div`
  display: grid;
  grid-template-columns: 10rem 10rem;
  padding: 1rem;
  justify-content: center;
  grid-column-gap: 1rem;
  align-items: center;
  height: 100%;
`
const StyledModal = styled(Modal)`
  @media only screen and (max-width: 680px) {
    width: 100%;
    border: none;
    margin: 0;
    position: absolute;
    bottom: 0;
    height: auto;
  }
`

export interface IInvoiceList {
  data: IInvoiceItem[]
  getInvoices: () => void,
  addInvoice: (v: IInvoiceItem) => void,
  editInvoice: (v1: IInvoiceItem, v2: IInvoiceItem) => void,
  deleteInvoice: (v: IInvoiceItem) => void,
  searchIBAN?: (iban: string) => void,
  selectItem: (v: IInvoiceItem) => void,
  setModeInvoice: (e: string, p?: IInvoiceItem) => void,
  resetState: () => void,
  setTabActive: (tabName: string) => void,
  setRetrieveFromBankAcc: (v: boolean) =>  void,
  selectPaymentItem: (v: IInvoiceItem) => void,
  isRetrieveFromBankAcc?: boolean,
  paymentsFromBankAccount: IInvoiceItem[],
  loading?: boolean,
  selectedItem: IInvoiceItem,
  modeInvoiceModal?: string,
  isDeleteMode? : boolean,
  activedTab?: string,
  selectedPaymentItem?: IInvoiceItem
}

const InvoiceList: React.FC<IInvoiceList> = (props) => {
  const {
    data,
    loading,
    selectItem,
    selectedItem,
    searchIBAN,
    getInvoices,
    addInvoice,
    editInvoice,
    deleteInvoice,
    paymentsFromBankAccount,
    modeInvoiceModal,
    isDeleteMode,
    setModeInvoice,
    resetState,
    setTabActive,
    activedTab,
    isRetrieveFromBankAcc,
    setRetrieveFromBankAcc,
    selectedPaymentItem,
    selectPaymentItem
  } = props;
  
  const [isOpenModal, setIsOpenModal] = useState(false);
  const isOpenModalAddAndEdit = isOpenModal && !isDeleteMode;
  const isOpenModalDelete = isOpenModal && !!isDeleteMode;

  useEffect(() => {
    getInvoices()
  }, [getInvoices])
  
  useEffect(() => {
    if (modeInvoiceModal) {
      setIsOpenModal(true)
    } else {
      setIsOpenModal(false)
    }
  }, [modeInvoiceModal])


  const onDelete = (value: IInvoiceItem) => setModeInvoice("DELETE", value)
  const onEdit = (value: IInvoiceItem) => setModeInvoice("EDIT", value)
  const onAdd = () => setModeInvoice("ADD")
  
  const onSubmit = (values) => {
    if (modeInvoiceModal === 'EDIT') {
      editInvoice(selectedItem, values)
    }

    if (modeInvoiceModal === 'DELETE') {
      deleteInvoice(selectedItem)
    }

    if (modeInvoiceModal === 'ADD') {
      addInvoice({ ...values, ...selectedItem })
    }
  }

  const onNext = (values) => {
    const valuesNew = { ...selectedItem, ...values }
    selectItem(valuesNew)
    setTabActive("pm")
  }

  const renderItem = (itemProps: IInvoiceItem, idx: number) => (
    <InvoiceItem
      {...itemProps}
      key={idx}
      onDelete={onDelete}
      onEdit={onEdit}
    />)

  return (
    <>
      <StyledInvoiceList>
        <StyledInvoiceListHead>
          <StyledInvoiceListTitle>
              <Text>Invoices List</Text>
            </StyledInvoiceListTitle>
          <StyledInvoiceListAddBtn>
            <Button color="primary" onClick={onAdd}> Add new invoice </Button>
          </StyledInvoiceListAddBtn>
        </StyledInvoiceListHead>
        <StyledInvoiceListBody>
          {
            loading
            ? <Loading />
            : <SimpleList data={data} renderItem={renderItem} />
          }
        </StyledInvoiceListBody>
      </StyledInvoiceList>
      <Modal title="Invoice Dialog" isOpen={isOpenModalAddAndEdit} onClose={resetState}>
        <Tabs defaultActivedTab={activedTab}>
            <TabItem title="Invoice Information" name="ii">
              <InvoiceInformation
                onSubmit={onSubmit}
                item={selectedItem}
                onRetriveFromBankAcc={setRetrieveFromBankAcc}
                onNext={onNext}
              />
            </TabItem>
            {
              isRetrieveFromBankAcc && 
              <TabItem title="Payments" name="pm">
                <InvoicePayment
                  data={paymentsFromBankAccount}
                  onSubmit={onSubmit}
                  onChangeIBANNum={searchIBAN}
                  onSelect={selectPaymentItem}
                  selected={selectedPaymentItem}
                />
              </TabItem>
            }
        </Tabs>
      </Modal>
      <StyledModal title="Do you want to delete this item ?" isOpen={isOpenModalDelete} onClose={resetState}>
        <StyledModalConfirm>
          <Button color="primary" onClick={onSubmit}>Yes</Button>
          <Button color="danger" onClick={resetState}>No</Button>
        </StyledModalConfirm>
      </StyledModal>
    </>
  )
}

export default InvoiceList
