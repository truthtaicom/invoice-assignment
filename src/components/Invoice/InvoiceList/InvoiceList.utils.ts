export const simpleInvoiceListData = [
  {
    date: '2018-01-01',
    subject: 'Rent january',
    amount: 500,
    unit: 'EUR',
    ibanNum: 'DE-12-1234-1234-12345-12',
  },
  {
    date: '2018-01-02',
    subject: 'Rent january',
    amount: 1500,
    unit: 'EUR',
    ibanNum: 'DE-12-1234-1234-12345-12',
  },
  {
    date: '2018-01-03',
    subject: 'Rent january',
    amount: 2500,
    unit: 'EUR',
    ibanNum: 'DE-12-1234-1234-12345-12',
  },
  {
    date: '2018-01-04',
    subject: 'Rent january',
    amount: 3500,
    unit: 'EUR',
    ibanNum: 'DE-12-1234-1234-12345-12',
  }
]

