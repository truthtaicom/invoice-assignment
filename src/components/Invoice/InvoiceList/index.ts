import InvoiceListContainer from './InvoiceList.container'

export { default as InvoiceItem } from './InvoiceItem'
export default InvoiceListContainer