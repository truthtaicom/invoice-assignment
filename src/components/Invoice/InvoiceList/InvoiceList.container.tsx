import { connect } from "react-redux";
import InvoiceList from "./InvoiceList";
import {
  getInvoices,
  addInvoice,
  searchIBAN,
  editInvoice,
  deleteInvoice,
  selectItem,
  setModeInvoice,
  resetState,
  setTabActive,
  setRetrieveFromBankAcc,
  selectPaymentItem
} from "../Invoice.actions";

const mapStateToProps = state => ({
  loading: state.invoiceReducer.request,
  data: state.invoiceReducer.data,
  error: state.invoiceReducer.error,
  paymentsFromBankAccount: state.invoiceReducer.paymentsFromBankAccount,
  selectedItem: state.invoiceReducer.selectedItem || {},
  modeInvoiceModal: state.invoiceReducer.modeInvoiceModal,
  isDeleteMode: state.invoiceReducer.isDeleteMode,
  activedTab: state.invoiceReducer.activedTab,
  isRetrieveFromBankAcc: state.invoiceReducer.isRetrieveFromBankAcc,
  selectedPaymentItem: state.invoiceReducer.selectedPaymentItem
});

const mapDispatchToProps = {
  getInvoices,
  addInvoice,
  editInvoice,
  deleteInvoice,
  selectItem,
  searchIBAN,
  setModeInvoice,
  resetState,
  setTabActive,
  setRetrieveFromBankAcc,
  selectPaymentItem
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(InvoiceList);
