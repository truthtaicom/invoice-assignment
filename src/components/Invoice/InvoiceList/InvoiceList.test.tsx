import React from 'react'
import { render } from 'react-testing-library'
import InvoiceList from './InvoiceList'

describe('<InvoiceList /> spec', () => {
  it('renders the component', () => {
    const component = render(
      <InvoiceList
        deleteInvoice={jest.fn}
        editInvoice={jest.fn}
        addInvoice={jest.fn}
        selectItem={jest.fn}
        data={[
          {
            date: '2018-01-01',
            subject: 'Rent january',
            amount: 500,
            unit: 'EUR',
            ibanNum: 'DE-12-1234-1234-12345-12',
          },
          {
            date: '2018-01-02',
            subject: 'Rent january',
            amount: 1500,
            unit: 'EUR',
            ibanNum: 'DE-12-1234-1234-12345-12',
          }
        ]}
        getInvoices={jest.fn}
        selectedItem={{}}
        setModeInvoice={jest.fn}
        setTabActive={jest.fn}
        resetState={jest.fn}
        setRetrieveFromBankAcc={jest.fn}
        selectPaymentItem={jest.fn}
      />
    )
    expect(component).toMatchSnapshot()
  })
});