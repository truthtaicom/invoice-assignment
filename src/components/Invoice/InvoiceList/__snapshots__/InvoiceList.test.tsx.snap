// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InvoiceList /> spec renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-cMljjf kNIQwG"
      >
        <div
          class="sc-jAaTju jnhfra"
        >
          <div
            class="sc-jDwBTQ dWtMUn"
          >
            <span
              class="sc-gzVnrw fHvJOv"
            >
              Invoices List
            </span>
          </div>
          <div
            class="sc-gPEVay bFVaHF"
          >
            <button
              class="sc-bdVaJa eMWQLK"
              color="primary"
            >
               Add new invoice 
            </button>
          </div>
        </div>
        <div
          class="sc-iRbamj bupFeE"
        >
          <div
            class="sc-htoDjs bxSFQy"
          >
            <div
              class="sc-kkGfuU bxPbff"
            >
              <span
                class="sc-gzVnrw fHvJOv"
              >
                2018-01-01
              </span>
              <span
                class="sc-gzVnrw fHvJOv"
              >
                Rent january
              </span>
              <span
                class="sc-gzVnrw fHvJOv"
              >
                500
                 
                EUR
              </span>
              <span
                class="sc-gzVnrw fHvJOv"
              >
                DE-12-1234-1234-12345-12
              </span>
              <div
                class="sc-iAyFgw jeUPJC"
              >
                <button
                  class="sc-bdVaJa eMWQLK"
                  color="primary"
                >
                  Edit
                </button>
                <button
                  class="sc-bdVaJa isSLun"
                  color="danger"
                >
                  Delete
                </button>
              </div>
            </div>
            <div
              class="sc-kkGfuU bxPbff"
            >
              <span
                class="sc-gzVnrw fHvJOv"
              >
                2018-01-02
              </span>
              <span
                class="sc-gzVnrw fHvJOv"
              >
                Rent january
              </span>
              <span
                class="sc-gzVnrw fHvJOv"
              >
                1500
                 
                EUR
              </span>
              <span
                class="sc-gzVnrw fHvJOv"
              >
                DE-12-1234-1234-12345-12
              </span>
              <div
                class="sc-iAyFgw jeUPJC"
              >
                <button
                  class="sc-bdVaJa eMWQLK"
                  color="primary"
                >
                  Edit
                </button>
                <button
                  class="sc-bdVaJa isSLun"
                  color="danger"
                >
                  Delete
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-cMljjf kNIQwG"
    >
      <div
        class="sc-jAaTju jnhfra"
      >
        <div
          class="sc-jDwBTQ dWtMUn"
        >
          <span
            class="sc-gzVnrw fHvJOv"
          >
            Invoices List
          </span>
        </div>
        <div
          class="sc-gPEVay bFVaHF"
        >
          <button
            class="sc-bdVaJa eMWQLK"
            color="primary"
          >
             Add new invoice 
          </button>
        </div>
      </div>
      <div
        class="sc-iRbamj bupFeE"
      >
        <div
          class="sc-htoDjs bxSFQy"
        >
          <div
            class="sc-kkGfuU bxPbff"
          >
            <span
              class="sc-gzVnrw fHvJOv"
            >
              2018-01-01
            </span>
            <span
              class="sc-gzVnrw fHvJOv"
            >
              Rent january
            </span>
            <span
              class="sc-gzVnrw fHvJOv"
            >
              500
               
              EUR
            </span>
            <span
              class="sc-gzVnrw fHvJOv"
            >
              DE-12-1234-1234-12345-12
            </span>
            <div
              class="sc-iAyFgw jeUPJC"
            >
              <button
                class="sc-bdVaJa eMWQLK"
                color="primary"
              >
                Edit
              </button>
              <button
                class="sc-bdVaJa isSLun"
                color="danger"
              >
                Delete
              </button>
            </div>
          </div>
          <div
            class="sc-kkGfuU bxPbff"
          >
            <span
              class="sc-gzVnrw fHvJOv"
            >
              2018-01-02
            </span>
            <span
              class="sc-gzVnrw fHvJOv"
            >
              Rent january
            </span>
            <span
              class="sc-gzVnrw fHvJOv"
            >
              1500
               
              EUR
            </span>
            <span
              class="sc-gzVnrw fHvJOv"
            >
              DE-12-1234-1234-12345-12
            </span>
            <div
              class="sc-iAyFgw jeUPJC"
            >
              <button
                class="sc-bdVaJa eMWQLK"
                color="primary"
              >
                Edit
              </button>
              <button
                class="sc-bdVaJa isSLun"
                color="danger"
              >
                Delete
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
